torchvision/models/swin_transformer.py

```diff
@@ -0,0 +1,516 @@
+        attn = attn.view(-1, num_heads, x.size(1), x.size(1))
+
+    attn = F.softmax(attn, dim=-1)
+    attn = F.dropout(attn, p=attention_dropout)
+
+    x = attn.matmul(v).transpose(1, 2).reshape(x.size(0), x.size(1), C)
+    x = F.linear(x, proj_weight, proj_bias)
+    x = F.dropout(x, p=dropout)
+
+    # reverse windows
+    x = x.view(B, pad_H // window_size, pad_W // window_size, window_size, window_size, C)
+    x = x.permute(0, 1, 3, 2, 4, 5).reshape(B, pad_H, pad_W, C)
+
+    # reverse cyclic shift
+    if shift_size > 0:
+        x = torch.roll(x, shifts=(shift_size, shift_size), dims=(1, 2))
+
+    # unpad features
+    x = x[:, :H, :W, :].contiguous()
+    return x
+
+
+torch.fx.wrap("shifted_window_attention")
+
+
+class ShiftedWindowAttention(nn.Module):
+    """
+    See :func:`shifted_window_attention`.
+    """
+
+    def __init__(
+        self,
+        dim: int,
+        window_size: int,
+        shift_size: int,
+        num_heads: int,
+        qkv_bias: bool = True,
+        proj_bias: bool = True,
+        attention_dropout: float = 0.0,
+        dropout: float = 0.0,
+    ):
+        super().__init__()
+        self.window_size = window_size
+        self.shift_size = shift_size
+        self.num_heads = num_heads
+        self.attention_dropout = attention_dropout
+        self.dropout = dropout
+        self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
+        self.proj = nn.Linear(dim, dim, bias=proj_bias)
+
+        # get pair-wise relative position index for each token inside the window
+        coords_h = torch.arange(self.window_size)
+        coords_w = torch.arange(self.window_size)
+        coords = torch.stack(torch.meshgrid(coords_h, coords_w, indexing="ij"))  # 2, Wh, Ww
+        coords_flatten = torch.flatten(coords, 1)  # 2, Wh*Ww
+        relative_coords = coords_flatten[:, :, None] - coords_flatten[:, None, :]  # 2, Wh*Ww, Wh*Ww
+        relative_coords = relative_coords.permute(1, 2, 0).contiguous()  # Wh*Ww, Wh*Ww, 2
+        relative_coords[:, :, 0] += self.window_size - 1  # shift to start from 0
+        relative_coords[:, :, 1] += self.window_size - 1
+        relative_coords[:, :, 0] *= 2 * self.window_size - 1
+        relative_position_index = relative_coords.sum(-1).view(-1)  # Wh*Ww*Wh*Ww
+
+        # define a parameter table of relative position bias
+        relative_position_bias_table = torch.zeros(
+            (2 * window_size - 1) * (2 * window_size - 1), num_heads
+        )  # 2*Wh-1 * 2*Ww-1, nH
+        nn.init.trunc_normal_(relative_position_bias_table, std=0.02)
+
+        relative_position_bias = relative_position_bias_table[relative_position_index]  # type: ignore[index]
+        relative_position_bias = relative_position_bias.view(
+            self.window_size * self.window_size, self.window_size * self.window_size, -1
+        )
+        self.relative_position_bias = nn.Parameter(relative_position_bias.permute(2, 0, 1).contiguous().unsqueeze(0))
```

Comment:
Unfortunately as @xiaohu2015 highlighted this increased the number of parameters. Proof:
```python
from torchvision import models as M
model = M.swin_t()
sum(p.numel() for p in model.parameters())
```

Joao's commit 41faba232668f7ac4273a0cf632c0d0130c7ce9c: 28596370
Hu's commit 2500ff3763146d57d959f358dad1a667c4db0db8: 28288354

@jdsgomes I think here we both missed the fact that `relative_position_index` contains repeated indexes. To verify checkout @xiaohu2015's code and check:
```python

>>> model.features[1][0].attn.relative_position_index
tensor([84, 83, 82,  ..., 86, 85, 84])
```

I think we should revert the change.

Reply:
> > @xiaohu2015 Hmmm... Did you try moving it on the constructor and got issues during training? Or that's something you didn't try and you are worried it might affect us?
> 
> I do a toy experiment, I found it works. But I am not sure if it will affect performance, maybe you can do end2end training to test it.

@datumbox @jdsgomes the toy experiment is not for swin_t, so I am sorry that I maybe give some incorrect information.

Reply:
the problem is that some compute can't be included in the `forward`

Reply:
No worries for the initial confusion with the toy example. I think you did awesome job flagging that we both missed your point. We can revert the change once Joao is back or feel free to revert at your end.

Reply:
I will revert this, thanks for pointing it out @xiaohu2015, and for reverting @datumbox